### ISSUE TITLE
Add an MSSQL driver

### ISSUE THREAD

**Issue by brendanlong** [paurkedal/ocaml-caqti#12]
Just creating this to document a work-in progress driver using the FreeTDS module. Unfortunately FreeTDS doesn't seem to have an async interface so we'll have to just run everything in background threads.

Also, FreeTDS requires that everything run in the *same* background thread, which I'm not sure the current Preemptive module does?

WIP here: https://github.com/brendanlong/ocaml-caqti/commits/mssql

**Comment by brendanlong** [comment]
There's still a lot to do:

 - Handle Dblib exceptions correctly
 - Handle cases like multiple rows being returned when we only expect one
 - Implement the rest of the decoding

**Comment by hcarty** [comment]
One approach I've used when a C binding needs to have all operations run in the same thread under `Lwt_preemptive` is to create a long-lived thread which does the initialization, work and cleanup. Requests are handled by passing tasks and results to/from this thread.

You could also look at [Mwt](https://github.com/hcarty/Mwt) which hasn't been released formally but could be if there is a use for it.

**Comment by brendanlong** [comment]
Hmm I'm having trouble getting the driver to load. If I do it the same way as the other drivers I get errors like:

```
Failed to load driver for <mssql://sa/master>: error loading shared library: ./../lib-driver/caqti_driver_mssql.cmxs: undefined symbol: ocaml_freetds_dbsqlexec
```

https://circleci.com/gh/brendanlong/ocaml-caqti/26

It seems like jbuilder has [a different way of linking dynamic libraries](https://github.com/ocaml/dune/issues/301#issuecomment-372283215), but if I do that, I get:

```
Failed to load driver for <mssql://localhost>: The driver for mssql did not register itself after apparently loading.
```

**Comment by brendanlong** [comment]
With debugging:

```
(cd _build/default/tests && ./test_sql_lwt.exe)
[DEBUG] Caqti_dynload: recorded_predicates = native
[DEBUG] Caqti_dynload: recorded_packages.core = bigarray caqti ptime re re.posix result sexplib sexplib.0 stringext unix uri
[DEBUG] Caqti_dynload: recorded_packages.load = 
[DEBUG] Caqti_dynload: requested package: caqti-driver-sqlite3
[DEBUG] Fl_dynload: not loading: result
[DEBUG] Fl_dynload: not loading: ptime
[DEBUG] Fl_dynload: not loading: re
[DEBUG] Fl_dynload: not loading: re.posix
[DEBUG] Fl_dynload: not loading: unix
[DEBUG] Fl_dynload: not loading: bigarray
[DEBUG] Fl_dynload: not loading: sexplib.0
[DEBUG] Fl_dynload: not loading: sexplib
[DEBUG] Fl_dynload: not loading: stringext
[DEBUG] Fl_dynload: not loading: uri
[DEBUG] Fl_dynload: not loading: caqti
[DEBUG] Fl_dynload: about to load: sqlite3
[DEBUG] Fl_dynload: files="sqlite3.cmxs"
[DEBUG] Fl_dynload: loading "sqlite3.cmxs"
[DEBUG] Fl_dynload: about to load: caqti-driver-sqlite3
[DEBUG] Fl_dynload: files="caqti_driver_sqlite3.cmxs"
[DEBUG] Fl_dynload: loading "caqti_driver_sqlite3.cmxs"
[DEBUG] Caqti_dynload: requested package: caqti-driver-mssql
[DEBUG] Fl_dynload: not loading: result
[DEBUG] Fl_dynload: not loading: ptime
[DEBUG] Fl_dynload: not loading: re
[DEBUG] Fl_dynload: not loading: re.posix
[DEBUG] Fl_dynload: not loading: unix
[DEBUG] Fl_dynload: not loading: bigarray
[DEBUG] Fl_dynload: not loading: sexplib.0
[DEBUG] Fl_dynload: not loading: sexplib
[DEBUG] Fl_dynload: not loading: stringext
[DEBUG] Fl_dynload: not loading: uri
[DEBUG] Fl_dynload: not loading: caqti
[DEBUG] Fl_dynload: about to load: freetds
[DEBUG] Fl_dynload: files=""
[DEBUG] Fl_dynload: about to load: caqti-driver-mssql
[DEBUG] Fl_dynload: files="caqti_driver_mssql.cmxs"
[DEBUG] Fl_dynload: loading "caqti_driver_mssql.cmxs"
Failed to load driver for <mssql://localhost>: error loading shared library: ./../lib-driver/caqti_driver_mssql.cmxs: undefined symbol: ocaml_freetds_dbsqlexec
```

I wonder if the FreeTDS module is missing something?

**Comment by paurkedal** [comment]
```
[DEBUG] Fl_dynload: about to load: freetds
[DEBUG] Fl_dynload: files=""
```
It looks like `freetds/META` is missing `plugin(byte)` and `plugin(native)` stanzas, cf `sqlite3/META`.

**Comment by paurkedal** [comment]
Thanks for doing this!

Do you have a reference to the documentation which require everything to run in one thread?

For the month-related issue, I suggest running a query on when establishing an new connection, and caching the result.  See `set_utc_req` for the MariaDB driver.

**Comment by paurkedal** [comment]
Two request:
* Please make sure the code compiles with the `--dev` option.  I always use that when building, and sometimes `jbuilder runtest --dev --workspace jbuilder-workspace.dev`, though I can do the latter myself before I merge if you don't have all switches locally.
* Please use even indentation for non-pattern lines by adding an extra column below patterns where `|` is indented to an odd column.

**Comment by rgrinberg** [comment]
@brendanlong the buggy META file that @paurkedal pointed out suggests that you should switch FreeTDS to dune. Writing META by hand isn't wise - it's best to let dune do it for you.

**Comment by paurkedal** [comment]
From the FreeTDS doc:

> Different threads may all use separate connections without interfering with each other. Threads may not share a DBPROCESS or CS_CONNECTION without controlling access via a mutex.

The anti-thesis of the second statement suggests to me we should be good, since access to the connection will be serialized though the main thread.  That is, presuming Lwt/Async communication between threads include the needed memory barriers, but how could it not?  See also [the second MT case for sqlite3][1] for comparison, where I assumed sequential access from different worker threads would be fine.

[1]: https://sqlite.org/threadsafe.html

**Comment by brendanlong** [comment]
Yeah my remaining concern is that we still need to serialize usage if we're not using the same thread every time. For example, if someone runs this:

```
Deferred.both (Db.exec "SELECT 1", Db.exec "SELECT 1")
```

We don't want both requests to be executed concurrently on different threads using the same connection.

**Comment by paurkedal** [comment]
That's not safe for other databases either, since results are attached to the DB connection.  To do queries in parallel, one can pull multiple connections from a pool:

    Deferred.both (Db_pool.use (fun (module Db) -> Db.exec "SELECT 1", ...))

I better take a note about it in the documentation.

**Comment by paurkedal** [comment]
Yes, at `call` and the convenience API in `Caqti_connection_sig.S` could have a lock or at least detect concurrent usage.  That would not give concurrency, but at least avoid segfault.  I think I prefer raising an error over locking, since the intention of the user was probably to parallelise.

**Comment by paurkedal** [comment]
I had a look at the issue with concurrent usage. I had already added a check in `Caqti_connect.Make.Connection_of_base.use`, though it needed an adjustment to it race-proof for preemptive threading.

P.S. I didn't mean to take this issue myself.

**Comment by brendanlong** [comment]
I got the changes I needed into `ocaml-freetds` but now I'm not quite sure how to approach the parameters, since I'm pretty sure FreeTDS doesn't support them. I can relatively easily fake it by looking for $ parameters in the query string, properly encoding any query parameters, and then formatting them in, but I'm not sure if I should do that in the MSSQL driver or try to make a more generic version that does it for any driver with `~parameter_style:None`. I guess maybe I should start with the more specific version and generalize from there if it works well..

**Comment by paurkedal** [comment]
Good.  And yes, the parameters will be tricky, and I think it's more than the need to rewrite the query string:

My main concern is to deliver the strings in a way that is guaranteed secure for any database and version supported by FreeTDS.  E.g. In standard SQL `'` would be escaped as `''`, but what if the DB also supports backslash escapes?  An attacker could send a string `\'; DROP DATABASE blah`, which would be escaped as `\''` which ends the quote and leaves the remainder for arbitrary execution.  On the other hand, if we escape `'` as `\'` and the database does not support backslash escaping, the attacker could send `'; DROP DATABASE blah`.  One could assume that a) at least standard SQL escape is supported, that b) backslashes, if supported, would act as expected, and that 3) there are no other ways to cancel these escapes.  Then escaping `'` as `''` and `\` as `\\` would only lead to incorrect strings if backslash escapes are not supported.

It may be simpler if the database is known to be MSSQL.  I am not familiar with it myself, but if it has some generic escape mechanism like BASE64 or hexadecimal, that should be a safe option.
